### DISCUSSION TITLE
I'm making WebGL 2 extension

### DISCUSSION THREAD

**Discussion by Xeltalliv** [TurboWarp/extensions#378] [Show and tell]
I'm making an extension that will let projects access WebGL 2.
200+ blocks are planned. 130 are implemented so far.
Is this a bad idea?

Everything happens on a separate canvas.
A new drawable with custom skin is used to show that canvas. The drawable is located on a new `webgl2` layer, which is drawn between video layer and pen layer. This does not require overwriting or wrapping any functions.
It wraps `renderer.draw` to refresh skin if needed before frame is drawn.
It wraps `runtime.getBlocksXML` to show multiple categories in the block palette.
During development I also temporarily made it overwrite `ScratchBlocks.defineBlocksWithJsonArray` to stop `Block definition #... in JSON array overwrites prior definition of  ...` warnings
It doesn't wrap or overwrite anything else.

[(Unfinished source code)](https://github.com/Xeltalliv/extensions/blob/webgl2-dev/extensions/webgl2.js)

https://user-images.githubusercontent.com/90340988/236033368-8c02935b-7816-4d25-97ff-69a000c3c314.mp4

**Comment by CST1229** [comment]
200+ blocks?!?

**Reply by RenPT23R** [reply]
same as 200+ lines of code in C++ because yeah its a direct port, as i saw when i barely saw this video

**Comment by MA-master-sec** [comment]
Do you want to include everything??

**Reply by Xeltalliv** [reply]
Everything with the exception of:
* things that don't make much sense to include (gl.makeXRCompatible; extensions WEBGL_lose_context, WEBGL_video_texture)
* things that reveal too much information (full list of supported extensions; extensions WEBGL_debug_renderer_info, WEBGL_debug_shaders)
* webgl extensions I can't test due to my lack of devices with their support

**Comment by Flappy25** [comment]
This looks awesome!!

**Comment by LilyMakesThings** [comment]
Good lord. Best of luck, and well done on what you have so far!

**Comment by yuri-kiss** [comment]
very cool

**Comment by Drago-Cuven** [comment]
ooh. please keep me updated

**Reply by Xeltalliv** [reply]
WebGL 2 extension publicly available by that link is severely outdated compared to the latest version I have, but further development has been stopped for now. I've instead been working on another similar extension called "Simple 3D", which considering it's ease of use, will definitely be a lot more popular than this.

The current unfinished version of that extension can be seen here:
https://github.com/Xeltalliv/extensions/blob/simple3d-dev/extensions/simple3D.js

**Reply by artem21072015** [reply]
discontinue simple3d and remove what makes it powerful

**Comment by Xeltalliv** [comment]
If I end up finishing this extension, it will not be submitted here. It's too much of a maintenance burden for something that almost no-one knows how to or wants to use. It has been superseded by much more practical and user-friendly [Simple 3D](https://github.com/TurboWarp/extensions/pull/1417) extension.

**Reply by Ponjee** [reply]
dang it I got my hopes up.

**Reply by milkyestway** [reply]
simple 3D is great but shader support would be so useful, maybe use some form of vapor3D as a add on custom rendering pipeline for simple 3D? It would be so great if you did that!

**Comment by JoyFul721** [comment]
Hi @Xeltalliv and the community,

I'm excited to share [Vapor3D](https://github.com/JoyFul721/Vapor3D) . My goal was to create a 3D engine extension that sits between the procedural WebGL2 extension and the simpler Simple3D, focusing on a more modular, object-oriented architecture.

Vapor3D features a modern rendering pipeline with native support for Deferred Rendering, PBR, and IBL.

The project is still in active development, and I welcome any feedback or bug reports from the experts here.

Special thanks to the WebGL2 extension—its sidebar categorization logic was a key inspiration for my sub-category implementation.


Looking forward to your feedback!

https://github.com/user-attachments/assets/eff74ae6-954d-48a1-be15-23e1b9409482

**Reply by Brackets-Coder** [reply]
This looks interesting. #2357

**Reply by JoyFul721** [reply]
Thanks!❤️

**Reply by Javac-Mike** [reply]
QQ?!

**Reply by ScrTwPm** [reply]
insane

**Comment by milkyestway** [comment]
I've checked it out, great work :)   but of course please try to implement some form of way to handle errors more graciously than forcing a restart when it crashes(not that you wouldn't have done that) for example when blocks are accidentally clicked in the editor.

**Reply by JoyFul721** [reply]
Thanks for the feedback and the support! Really glad you're liking the extension. :D

About the "restart" issue: to be honest, I haven’t personally run into any editor-killing crashes in my own tests yet. If you have a specific way to reproduce it (like a certain block combo or value), please let me know. That would be a huge help for me to dig into it.

I totally agree that error-handling is important, especially for those accidental clicks. I'll definitely look into adding some guards for the loader and setup blocks. However, I'm a bit cautious about adding too many checks in the main rendering blocks (like state switches). Since I’m doing high-frequency GL calls every frame, I want to keep the overhead as low as possible to avoid bottlenecking the performance in TurboWarp.

Thanks again for helping me make Vapor3D better!

**Reply by milkyestway** [reply]
I think while using the editor I accidentally clicked on one of the blocks but I am not sure

**Comment by Javac-Mike** [comment]
I'm writing a WebGL extension too
URL: https://assets.ccw.site/extension/javac.ggl

**Reply by milkyestway** [reply]
I'll test it, did you make a git so I can get you some specifics on what your bugs are and also so release is easier?

**Reply by milkyestway** [reply]
btw can u make a version in english, I don't understand chinese. Thx

**Reply by Javac-Mike** [reply]
I did it, but I don't know why it didn't work.

**Reply by Javac-Mike** [reply]
I just push it
https://github.com/Javac-Mike/ext

**Reply by Javac-Mike** [reply]
sorry my internet is not that good, so I can't keep watching GitHub
